epsagon/lambda_trigger.go

```diff
@@ -43,50 +35,106 @@ func mapParametersToString(params map[string]string) string {
+		Type:    "trigger-creation",
+		Message: fmt.Sprintf("failed to convert rawEvent to %s. %v", assertedType, rawEvent),
+		Time:    tracer.GetTimestamp(),
 	}
-	triggerEvent := &protocol.Event{
-		Id:        event.RequestContext.RequestID,
+}
+
+func getAPIGatewayTriggerEvent(eventFields *APIGatewayEventFields, metadataOnly bool) *protocol.Event {
+	triggerEvent := getAPIGatewayBaseEvent(eventFields)
+	if !metadataOnly {
+		addAPIGatewayEventMetadata(triggerEvent, eventFields)
+	}
+	return triggerEvent
+}
+
+func getAPIGatewayBaseEvent(eventFields *APIGatewayEventFields) *protocol.Event {
+	return &protocol.Event{
+		Id:        eventFields.requestId,
 		Origin:    "trigger",
 		StartTime: tracer.GetTimestamp(),
 		Resource: &protocol.Resource{
-			Name:      event.Headers["Host"],
+			Name:      eventFields.host,
 			Type:      "api_gateway",
-			Operation: event.HTTPMethod,
+			Operation: eventFields.httpMethod,
 			Metadata: map[string]string{
-				"stage":                   event.RequestContext.Stage,
-				"query_string_parameters": mapParametersToString(event.QueryStringParameters),
-				"path_parameters":         mapParametersToString(event.PathParameters),
-				"path":                    event.Resource,
+				"stage":                   eventFields.stage,
+				"query_string_parameters": mapParametersToString(eventFields.queryStringParameters),
+				"path_parameters":         mapParametersToString(eventFields.pathParameters),
+				"path":                    eventFields.path,
 			},
 		},
 	}
-	if !metadataOnly {
-		if bodyJSON, err := json.Marshal(event.Body); err != nil {
-			tracer.AddException(&protocol.Exception{
-				Type:      "trigger-creation",
-				Message:   fmt.Sprintf("Failed to serialize body %s", event.Body),
-				Traceback: string(debug.Stack()),
-				Time:      tracer.GetTimestamp(),
-			})
-			triggerEvent.Resource.Metadata["body"] = ""
-		} else {
-			triggerEvent.Resource.Metadata["body"] = string(bodyJSON)
-		}
-		triggerEvent.Resource.Metadata["headers"] = mapParametersToString(event.Headers)
+}
+
+func addAPIGatewayEventMetadata(triggerEvent *protocol.Event, eventFields *APIGatewayEventFields) {
+	if bodyJSON, err := json.Marshal(eventFields.body); err != nil {
+		tracer.AddException(&protocol.Exception{
+			Type:      "trigger-creation",
+			Message:   fmt.Sprintf("Failed to serialize body %s", eventFields.body),
+			Traceback: string(debug.Stack()),
+			Time:      tracer.GetTimestamp(),
+		})
+		triggerEvent.Resource.Metadata["body"] = ""
+	} else {
+		triggerEvent.Resource.Metadata["body"] = string(bodyJSON)
 	}
+	triggerEvent.Resource.Metadata["headers"] = mapParametersToString(eventFields.headers)
+}
 
-	return triggerEvent
+func triggerAPIGatewayProxyRequest(rawEvent interface{}, metadataOnly bool) *protocol.Event {
+	event, ok := rawEvent.(lambdaEvents.APIGatewayProxyRequest)
+	if !ok {
+		assertionException := getEventAssertionException(rawEvent, "lambdaEvents.APIGatewayProxyRequest")
```

Comment:
is there a check for the returned event value? (to make sure the code won't panic when the func returns nil)

Reply:
Yes, the `addLambdaTrigger` function that initiates the trigger event creation flow ignores `nil` events.
(line 495, lambda_trigger.go)
```go
func addLambdaTrigger(
	payload json.RawMessage,
	metadataOnly bool,
	triggerFactories map[string]factoryAndType,
	currentTracer tracer.Tracer,
) {
	var triggerEvent *protocol.Event

	triggerSource := guessTriggerSource(payload)

	if triggerSource == "json" {
		triggerEvent = triggerJSONEvent(payload, metadataOnly)
	} else if triggerSource == "api_gateway_no_proxy" {
		// currently not supported, needs to extract data from json
	} else {
		factoryStruct, found := triggerFactories[triggerSource]
		if found {
			triggerEvent = decodeAndUnpackEvent(
				payload, factoryStruct.EventType, factoryStruct.Factory, metadataOnly)
		}
	}

	// If a trigger was found
	if triggerEvent != nil {
		currentTracer.AddEvent(triggerEvent)
	}
}

```